pip/pip-311.md

```diff
@@ -0,0 +1,73 @@
+# Title: Add a request ID parameter to the `newMultiTransactionMessageAck` method signature
+
+# Background knowledge
+**newMultiTransactionMessageAck**: A public API in `Commands.java` that is used to assemble a `CommandAck` to individually ack a list of message IDs with a transaction ID.
+
+
+# Motivation
+In Pulsar, all operations are encapsulated as individual commands.
+For instance, the action of sending a message by ClientCnx is a `Send` command, and the broker has a corresponding `handleSend` method to handle the request of sending messages. 
+Subsequently, the broker sends a successful command, and ClientCnx will have a corresponding 'HandleSuccess' method to process the returned successful command.
+ClientCnx locally caches each request's ID and future. Upon receiving a response from the broker, it operates on the corresponding future based on the request ID. 
+For requests that do not receive a response within a given timeout, their futures will be completed with a Timeout exception and removed from the local cache map.
+This requires that all command requests waiting for a response should include the request ID, so that the response returned by the broker can carry the corresponding request ID. 
+This way, ClientCnx can clearly identify which request has been processed and completed.
+
+But the public API `newMultiTransactionMessageAck` in the `Commands.java` creates the ack commands without request ID and hopes to wait for the response.
+This causes all the requests created by `newMultiTransactionMessageAck` to be timed out.
```

Comment:
From the source code, I don't see any places that used `newMultiTransactionMessageAck`.  What is the actual problem that the proposal want to fix? 

And for the transaction ack, we are using the following method to create the ack command, right?

```java
    public static ByteBuf newAck(long consumerId, List<MessageIdData> messageIds, AckType ackType,
                                 ValidationError validationError, Map<String, Long> properties, long txnIdLeastBits,
                                 long txnIdMostBits, long requestId) {
        BaseCommand cmd = localCmd(Type.ACK);
        CommandAck ack = cmd.setAck()
                .setConsumerId(consumerId)
                .setAckType(ackType);
        ack.addAllMessageIds(messageIds);

        return newAck(validationError, properties, txnIdLeastBits, txnIdMostBits, requestId, ack, cmd);
    }
```

Reply:
+1

Reply:
For the chunk messages, the transaction ack should use this command. The current transaction ack for the chunk messages is wrong,  I fix it in https://github.com/apache/pulsar/pull/21268.

Reply:
So there are two problems:
 1. The transactional ack for chunk messages should use `newMultiTransactionMessageAck `.
 2. The `newMultiTransactionMessageAck` does not pass the request ID, and the transaction ack op needs to wait for a response.
 This proposal resolves the problem 2. 
 
 In addition, even if `newMultiTransactionMessageAck` is not used temporarily, we should ensure its correctness.

Reply:
@liangyepianzhou Could you please point out why this [method](https://github.com/apache/pulsar/blob/c6704dfcd977c790168e4bbad36ac67b555a3041/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java#L1038-L1040) can't handle the chunk message ack?

Now we have 8 methods to create ack command. We should consider to check which one is redundant.

<img width="638" alt="image" src="https://github.com/apache/pulsar/assets/12592133/a833955f-e1e5-4df9-bb7e-f9cb78917239">

Reply:
>@liangyepianzhou Could you please point out why this [method](https://github.com/apache/pulsar/blob/c6704dfcd977c790168e4bbad36ac67b555a3041/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java#L1038-L1040) can't handle the chunk message ack?

If the messageId is a `ChunkMessageIdImpl`, the ledger ID and entry ID will belong to the `lastChunkMsgId`.
There will be ack holes for the individual ack of chunk messages.
https://github.com/apache/pulsar/blob/2b5c199053a5b2d7f849e6604d619bae9197a8c9/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L2791-L2814

https://github.com/apache/pulsar/blob/2b5c199053a5b2d7f849e6604d619bae9197a8c9/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ChunkMessageIdImpl.java#L30-L33

Reply:
>Now we have 8 methods to create ack command. We should consider to check which one is redundant.

IMO, just one public method to create ack command is enough. Because the `message_id` is repeated and the `TxnID` is optional. Do we need to correct it?
```
    public static ByteBuf newMultiMessageAck(long consumerId, TxnID txnID,
            List<Triple<Long, Long, ConcurrentBitSetRecyclable>> entries, long requestID) 
```